### PR TITLE
Adds Solr operator.

### PR DESCRIPTION
**Pull request by nicksantamaria**
High-level overview of changes included:

* Adds new controller CLI command `skpr-operator search solr` with flags to configure various behaviors of the operator.
* Adds new CRD for a solr service which allows configuration of the solr version and resource sizing (`small`, `medium` and `large` available).

Requesting a `Solr` results in creation of the following resources:

* `StatefulSet` which handles deployment of the solr pod.
* `PeristentVolumeClaim` created by the statefulset for the data volume.
* `Service` which allows apps to connect to the solr service.

### CODE REVIEW

**Comment by nicksantamaria** [comment]
Shout out to @acbramley for his work in previousnext/k8s-solr which was ripped off massively for this one.

File: internal/controller/search/solr/controller.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+// Add creates a new Solr Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, coreName, dataVolumePath, image, imageInit, imageUser, storageClass string, port int) error {
+	return add(mgr, newReconciler(mgr, coreName, dataVolumePath, image, imageInit, imageUser, storageClass, port))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, coreName, dataVolumePath, image, imageInit, imageUser, storageClass string, port int) reconcile.Reconciler {
+	return &ReconcileSolr{
+		coreName:       coreName,
```
Comment (nickschuch) [inline internal/controller/search/solr/controller.go]: Make a `Params` struct with these in it.

File: internal/controller/search/solr/controller.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				corev1.ServicePort{
+					Port: int32(r.port),
+				},
+			},
+			Selector: map[string]string{
+				"addon": name,
+			},
+			Type: "ClusterIP",
```
Comment (nickschuch) [inline internal/controller/search/solr/controller.go]: Checkout `k8s.io/api/core/v1` for the ClusterIP const

File: internal/controller/search/solr/controller.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+							Name:  "solr",
+							Image: fmt.Sprintf("%s:%s", r.image, solr.Spec.Version),
+							Ports: []corev1.ContainerPort{
+								corev1.ContainerPort{
+									ContainerPort: int32(r.port),
+								},
+							},
+							Env: []corev1.EnvVar{
+								corev1.EnvVar{
+									Name:  "SOLR_HEAP",
```
Comment (nickschuch) [inline internal/controller/search/solr/controller.go]: Const?

File: internal/controller/search/solr/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
+package solr
+
+import (
+	"testing"
+	"time"
+
+	"github.com/onsi/gomega"
+	searchv1beta1 "github.com/skpr/operator/pkg/apis/search/v1beta1"
```
Comment (nickschuch) [inline internal/controller/search/solr/controller_test.go]: New group

File: internal/controller/search/solr/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
+package solr
```
Comment (nickschuch) [inline internal/controller/search/solr/controller_test.go]: Time to write a unit test ;)

File: internal/controller/search/solr/size_profiles.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package solr
```
Comment (nickschuch) [inline internal/controller/search/solr/size_profiles.go]: Extract these out into Size objects? :D Checkout the workflow project.

File: pkg/apis/search/v1beta1/solr_types.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	SolrSizeSmall SolrSize = "small"
+	// SolrSizeMedium refers to the medium size profile.
+	SolrSizeMedium SolrSize = "medium"
+	// SolrSizeLarge refers to the large size profile.
+	SolrSizeLarge SolrSize = "large"
+)
+
+// SolrSpec defines the desired state of Solr
+type SolrSpec struct {
+	Size    SolrSize `json:"size"`
```
Comment (nickschuch) [inline pkg/apis/search/v1beta1/solr_types.go]: Comments!

File: pkg/apis/search/v1beta1/solr_types.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	Version string   `json:"version"`
+}
+
+// SolrStatus defines the observed state of Solr
+type SolrStatus struct{}
```
Comment (nickschuch) [inline pkg/apis/search/v1beta1/solr_types.go]: Status should expose some of the connection details eg. Service IP / Core / Port.

**Comment by nickschuch** [comment]
I think we should integrate this one with the `Drupal` object as well.

**Comment by nicksantamaria** [comment]
Related: https://github.com/skpr/containers/pull/15

File: cmd/skpr-operator/search/solr.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+		Envar("SKPR_SEARCH_SOLR_IMAGE_USER").
+		Default(defaultImageUser).
+		StringVar(&c.ImageUser)
+	cmd.Flag("storage-class", "The storage class to use for solr data volume").
+		Envar("SKPR_SEARCH_SOLR_STORAGE_CLASS").
+		StringVar(&c.StorageClass)
+	cmd.Flag("port", "Container port to expose").
+		Envar("SKPR_SEARCH_SOLR_PORT").
+		Default(defaultPort).
+		IntVar(&c.Port)
```
Comment (acbramley) [inline cmd/skpr-operator/search/solr.go]: None of our other commands are indented like this, do we want to standardise on one way or another?

Comment (acbramley) [reply cmd/skpr-operator/search/solr.go]: FWIW I quite like it like this.

File: internal/controller/search/solr/controller.go [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+	}
+
+	// Watch for changes to Solr
+	err = c.Watch(&source.Kind{Type: &searchv1beta1.Solr{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// TODO(user): Modify this to be the types you create
+	// Uncomment watch a Deployment created by Solr - change this for objects you create
```
Comment (acbramley) [inline internal/controller/search/solr/controller.go]: Left over generated comments

Comment (acbramley) [reply internal/controller/search/solr/controller.go]: Infact, are we creating a deployment here?

File: internal/controller/search/solr/controller.go [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+
+// Reconcile reads that state of the cluster for a Solr object and makes changes based on the state read
+// and what is in the Solr.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  The scaffolding writes
+// a Deployment as an example
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=statefulsets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=services/status,verbs=get;update;patch
```
Comment (acbramley) [inline internal/controller/search/solr/controller.go]: Our other operators don't list all the objects they are managing, do we need to update them?

File: internal/controller/search/solr/controller.go [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+
+	client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// Reconcile reads that state of the cluster for a Solr object and makes changes based on the state read
+// and what is in the Solr.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  The scaffolding writes
+// a Deployment as an example
```
Comment (acbramley) [inline internal/controller/search/solr/controller.go]: Left over autogenerated comment

File: internal/controller/search/solr/controller.go [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+					},
+					Namespace: solr.ObjectMeta.Namespace,
+				},
+				Spec: corev1.PodSpec{
+					TerminationGracePeriodSeconds: &timeoutGracePeriod,
+					InitContainers: []corev1.Container{
+						// Our Solr containers run as the user "solr".
+						// This container will ensure that the permissions are set.
+						// Otherwise Solr will fail to boot in the first instance.
+						corev1.Container{
```
Comment (acbramley) [inline internal/controller/search/solr/controller.go]: This is a redundant type declaration, there are a bunch more in this file that can be removed. GoLand highlights these for you ;)

File: pkg/apis/search/v1beta1/v1beta1_suite_test.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
+package v1beta1
```
Comment (acbramley) [inline pkg/apis/search/v1beta1/v1beta1_suite_test.go]: This file can goooo

File: pkg/apis/search/v1beta1/solr_types_test.go [hunk 15]
```diff
@@ -0,0 +1,42 @@
+package v1beta1
```
Comment (acbramley) [inline pkg/apis/search/v1beta1/solr_types_test.go]: Needs a `// +build integration`

File: cmd/skpr-operator/search/solr.go (hunk 9 above)
Comment (nicksantamaria) [reply cmd/skpr-operator/search/solr.go]: I think we should. The single line syntax is annoying to scan.

File: internal/controller/search/solr/controller.go (hunk 10 above)
Comment (nicksantamaria) [reply internal/controller/search/solr/controller.go]: No this one is creating a statefulset. Its similar to a deployment but better suited to databases.

File: internal/controller/search/solr/controller.go (hunk 11 above)
Comment (nicksantamaria) [reply internal/controller/search/solr/controller.go]: The other operators should have these annotations for things they manage. Not an issue for kubedev as everything runs as admin, but without the right rbac it wont work in prod.

File: internal/controller/search/solr/size_profiles.go (hunk 6 above)
Comment (nicksantamaria) [reply internal/controller/search/solr/size_profiles.go]: Do we need this feature in this PR? Just being mindful of remaining budget.

File: pkg/apis/search/v1beta1/solr_types.go (hunk 8 above)
Comment (nicksantamaria) [reply pkg/apis/search/v1beta1/solr_types.go]:
```
$ kubectl get solr dev -o yaml

apiVersion: search.skpr.io/v1beta1
kind: Solr
metadata:
  name: dev
  namespace: default
spec:
  size: small
  version: "7"
status:
  core_name: core1
  service_host: 10.43.136.135
  service_port: 8983
```

File: internal/controller/search/solr/controller.go (hunk 13 above)
Comment (nicksantamaria) [reply internal/controller/search/solr/controller.go]: Thanks for the tip - definitely happy to be back in GoLand.

File: cmd/skpr-operator/search/solr.go (hunk 9 above)
Comment (acbramley) [reply cmd/skpr-operator/search/solr.go]: 100% agree

File: internal/controller/search/solr/controller.go (hunk 11 above)
Comment (acbramley) [reply internal/controller/search/solr/controller.go]: We'll need to do a big parse for that then cause most of them are missing

File: internal/controller/search/solr/controller_test.go (hunk 5 above)
Comment (nicksantamaria) [reply internal/controller/search/solr/controller_test.go]: doneskies!

File: cmd/skpr-operator/search/solr.go (hunk 9 above)
Comment (nicksantamaria) [reply cmd/skpr-operator/search/solr.go]: https://github.com/skpr/operator/pull/48

File: pkg/apis/search/v1beta1/v1beta1_suite_test.go [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"log"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
```
Comment (nickschuch) [inline pkg/apis/search/v1beta1/v1beta1_suite_test.go]: Set this to https://github.com/nickschuch/controller-runtime-envtest-k3s

**Pull request merged**